connectors/sources/onedrive.py

```diff
@@ -0,0 +1,219 @@
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def _set_access_token(self):
+        """Generate access token with configuration fields and stores it in the cache"""
+        url = f"https://login.microsoftonline.com/{self.tenant_id}/oauth2/v2.0/token"
+        data = {
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+            "grant_type": "client_credentials",
+            "scope": "https://graph.microsoft.com/.default",
+        }
+        headers = {"Content-Type": "application/x-www-form-urlencoded"}
+        async with aiohttp.request(
+            method="POST", url=url, data=data, headers=headers, raise_for_status=True
+        ) as response:
+            token_reponse = await response.json()
+            self.access_token = token_reponse["access_token"]
+            self.token_expires_at = datetime.utcnow() + timedelta(
+                seconds=int(token_reponse.get("expires_in", 0))
+            )
+            self._token_cache.set(self.access_token, self.token_expires_at)
+
+
+class OneDriveClient:
+    """Client Class for API calls to OneDrive"""
+
+    def __init__(self, configuration):
+        self._sleeps = CancellableSleeps()
+        self.configuration = configuration
+        self.retry_count = self.configuration["retry_count"]
+        self._logger = logger
+        self.token = AccessToken(configuration=configuration)
+
+    def set_logger(self, logger_):
+        self._logger = logger_
+
+    @cached_property
+    def _get_session(self):
+        """Generate base client session with configuration fields
+        Returns:
+            ClientSession: Base client session
+        """
+
+        timeout = aiohttp.ClientTimeout(total=None)
+        return aiohttp.ClientSession(
+            timeout=timeout,
+            raise_for_status=True,
+        )
+
+    async def close_session(self):
+        self._sleeps.cancel()
+        await self._get_session.close()
+        del self._get_session
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def api_call(self, url_name):
+        url = parse.urljoin(BASE_URL, url_name)
+        access_token = await self.token.get()
+        headers = {"authorization": f"Bearer {access_token}"}
+        async with self._get_session.get(url=url, headers=headers) as response:
+            yield response
+
+
+class OneDriveDataSource(BaseDataSource):
+    """OneDrive"""
+
+    name = "OneDrive"
+    service_type = "onedrive"
+
+    def __init__(self, configuration):
+        """Setup the connection to OneDrive
+
+        Args:
+            configuration (DataSourceConfiguration): Object of DataSourceConfiguration class.
+        """
+        super().__init__(configuration=configuration)
+        self.onedrive_client = OneDriveClient(configuration)
```

Comment:
Should we defer the initilization of the client by using a getter decorated with `@cached_property`? From a lifecycle perspective configuration validation is executed before `ping`.